File: cwe_comparative/CWE-22_path_traversal/vulnerable.go
// CWE-22: Path Traversal (Go)
// filepath.Join resolves ".." — it does NOT strip traversal sequences
package main

import (
	"fmt"
	"os"
	"path/filepath"
)

const baseDir = "/var/www/files"

func readFile(userFilename string) {
	// BUG: filepath.Join resolves ".." components
	// filepath.Join("/var/www/files", "../../etc/passwd") = "/etc/passwd"
	fullPath := filepath.Join(baseDir, userFilename)

	data, err := os.ReadFile(fullPath)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Println(string(data))
}

func main() {
	readFile("../../etc/passwd")
}

File: cwe_comparative/CWE-22_path_traversal/vulnerable.rs
// CWE-22: Path Traversal (Rust)
// Path::join does NOT sanitize ".." — it resolves components as-is
use std::fs;
use std::path::Path;

const BASE_DIR: &str = "/var/www/files";

fn read_file(user_filename: &str) {
    // BUG: join preserves ".." — path escapes the base directory
    // Path::new("/var/www/files").join("../../etc/passwd") = "/etc/passwd"
    let full_path = Path::new(BASE_DIR).join(user_filename);

    match fs::read_to_string(&full_path) {
        Ok(content) => println!("{}", content),
        Err(e) => println!("Error: {}", e),
    }
}

fn main() {
    read_file("../../etc/passwd");
}
